File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/meeting/StateMeeting.scala
package ch.epfl.pop.model.network.method.message.data.meeting

import ch.epfl.pop.json.MessageDataProtocol._
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.data.ActionType.ActionType
import ch.epfl.pop.model.network.method.message.data.ObjectType.ObjectType
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.{Hash, Timestamp, WitnessSignaturePair}
import spray.json._

final case class StateMeeting(
    id: Hash,
    name: String,
    creation: Timestamp,
    last_modified: Timestamp,
    location: Option[String],
    start: Timestamp,
    end: Option[Timestamp],
    extra: Option[Any],
    modification_id: Hash,
    modification_signatures: List[WitnessSignaturePair]
) extends MessageData {
  override val _object: ObjectType = ObjectType.MEETING
  override val action: ActionType = ActionType.STATE
}

object StateMeeting extends Parsable {
  def apply(
      id: Hash,
      name: String,
      creation: Timestamp,
      last_modified: Timestamp,
      location: Option[String],
      start: Timestamp,
      end: Option[Timestamp],
      extra: Option[Any],
      modification_id: Hash,
      modification_signatures: List[WitnessSignaturePair]
  ): StateMeeting = {
    new StateMeeting(id, name, creation, last_modified, location, start, end, extra, modification_id, modification_signatures)
  }

  override def buildFromJson(payload: String): StateMeeting = payload.parseJson.asJsObject.convertTo[StateMeeting]
}

File: fe1-web/src/features/meeting/network/messages/StateMeeting.ts
import { ActionType, MessageData, ObjectType } from 'core/network/jsonrpc/messages';
import { validateDataObject } from 'core/network/validation';
import { checkTimestampStaleness, checkWitnessSignatures } from 'core/network/validation/Checker';
import {
  Hash,
  ProtocolError,
  PublicKey,
  Signature,
  Timestamp,
  WitnessSignature,
} from 'core/objects';

/** Data received to track the state of a Meeting */
export class StateMeeting implements MessageData {
  public readonly object: ObjectType = ObjectType.MEETING;

  public readonly action: ActionType = ActionType.STATE;

  public readonly id: Hash;

  public readonly name: string;

  public readonly creation: Timestamp;

  public readonly last_modified: Timestamp;

  public readonly location?: string;

  public readonly start: Timestamp;

  public readonly end?: Timestamp;

  public readonly extra?: {};

  public readonly modification_id: Hash;

  public readonly modification_signatures: WitnessSignature[];

  constructor(msg: Partial<StateMeeting>) {
    if (!msg.name) {
      throw new ProtocolError("Undefined 'name' parameter encountered during 'StateMeeting'");
    }
    this.name = msg.name;

    if (!msg.creation) {
      throw new ProtocolError("Undefined 'creation' parameter encountered during 'StateMeeting'");
    }
    checkTimestampStaleness(msg.creation);
    this.creation = msg.creation;

    if (!msg.last_modified) {
      throw new ProtocolError(
        "Undefined 'last_modified' parameter encountered during 'StateMeeting'",
      );
    }
    if (msg.last_modified < msg.creation) {
      throw new ProtocolError(
        "Invalid timestamp encountered: 'last_modified' parameter smaller than 'creation'",
      );
    }
    this.last_modified = msg.last_modified;

    if (msg.location) {
      this.location = msg.location;
    }

    if (!msg.start) {
      throw new ProtocolError("Undefined 'start' parameter encountered during 'StateMeeting'");
    }
    checkTimestampStaleness(msg.start);
    this.start = msg.start;

    if (msg.end) {
      if (msg.end < msg.creation) {
        throw new ProtocolError(
          "Invalid timestamp encountered: 'end' parameter smaller than 'creation'",
        );
      }
      if (msg.end < msg.start) {
        throw new ProtocolError(
          "Invalid timestamp encountered: 'end' parameter smaller than 'start'",
        );
      }
      this.end = msg.end;
    }

    if (msg.extra) {
      this.extra = JSON.parse(JSON.stringify(msg.extra));
    } // clone JS object extra

    if (!msg.modification_id) {
      throw new ProtocolError(
        "Undefined 'modification_id' parameter encountered during 'StateMeeting'",
      );
    }
    // TODO: checkModificationId(msg.modification_id);
    this.modification_id = msg.modification_id;

    if (!msg.modification_signatures) {
      throw new ProtocolError(
        "Undefined 'modification_signatures' parameter encountered during 'StateMeeting'",
      );
    }
    checkWitnessSignatures(msg.modification_signatures, msg.modification_id);
    this.modification_signatures = [...msg.modification_signatures];

    if (!msg.id) {
      throw new ProtocolError("Undefined 'id' parameter encountered during 'StateMeeting'");
    }

    // FIXME: implementation not finished, get event from storage,
    /*
    const lao: Lao = OpenedLaoStore.get();
    const expectedHash = Hash.fromArray(
      EventTags.MEETING, lao.id, lao.creation, MEETING_NAME,
    );
    if (!expectedHash.equals(msg.id))
      throw new ProtocolError(
        'Invalid \'id\' parameter encountered during \'StateMeeting\': unexpected id value'
      ); */
    this.id = msg.id;
  }

  /**
   * Creates a StateMeeting object from a given object.
   *
   * @param obj
   */
  public static fromJson(obj: any): StateMeeting {
    const { errors } = validateDataObject(ObjectType.MEETING, ActionType.STATE, obj);

    if (errors !== null) {
      throw new ProtocolError(`Invalid meeting state\n\n${errors}`);
    }

    return new StateMeeting({
      ...obj,
      creation: new Timestamp(obj.creation),
      last_modified: new Timestamp(obj.last_modified),
      start: new Timestamp(obj.start),
      end: obj.end !== undefined ? new Timestamp(obj.end) : undefined,
      modification_id: new Hash(obj.modification_id),
      modification_signatures: obj.modification_signatures.map(
        (ws: any) =>
          new WitnessSignature({
            witness: new PublicKey(ws.witness),
            signature: new Signature(ws.signature),
          }),
      ),
      id: new Hash(obj.id),
    });
  }
}
